File: kompartment/src/domain/distribution.js
/**
 * What a sample of one output looks like, in numbers.
 *
 * A band on a chart is the 5th and 95th percentile at every time. Asked "what
 * is the probability the dose exceeds the limit", or "how sure is that mean",
 * the band has no answer, and those are the questions a probabilistic
 * assessment exists to put numbers on. GoldSim's Distribution Summary is the
 * page that does: mean with confidence bounds, standard deviation, skewness,
 * kurtosis, the count, a percentile table, a calculator between value and
 * probability, and a conditional tail expectation. This is that page's
 * arithmetic, over one column of the matrix a run keeps -- one output at one
 * time -- sorted once so that every question after the first is a lookup.
 *
 * Everything here is the plain sample statistic, with the one convention
 * stated: the percentile at `p` is the value at position `p·(n−1)` of the
 * sorted sample, linearly interpolated, which is what the bands use rounded
 * and what a spreadsheet's PERCENTILE gives. The confidence bounds are the
 * ones a finite sample honestly supports -- normal-approximation bounds on the
 * mean, and the Dvoretzky–Kiefer–Wolfowitz band on the whole distribution
 * function, which holds for any shape at all and says how much of what the
 * CDF shows is sampling noise.
 */

/**
 * One output at one time, over the realisations, sorted and finite.
 *
 * @param {Float64Array} values  realisation-major, `iterations × times`
 * @param {Uint8Array|null} [mask]  1 to keep a realisation, 0 to screen it out
 */
export function sortedColumn(values, times, iterations, at, mask = null) {
	const out = new Float64Array(iterations);
	let n = 0;
	for (let i = 0; i < iterations; i++) {
		if (mask && !mask[i]) continue;
		const v = values[i * times + at];
		if (Number.isFinite(v)) out[n++] = v;
	}
	return out.subarray(0, n).sort();
}

/** The value at cumulative probability `p`, interpolated between order statistics. */
export function valueAt(sorted, p) {
	const n = sorted.length;
	if (!n) return NaN;
	if (n === 1) return sorted[0];
	const x = Math.min(1, Math.max(0, p)) * (n - 1);
	const lo = Math.floor(x);
	const hi = Math.min(n - 1, lo + 1);
	return sorted[lo] + (sorted[hi] - sorted[lo]) * (x - lo);
}

/** The fraction of the sample at or below `x`: the empirical CDF. */
export function probabilityOf(sorted, x) {
	const n = sorted.length;
	if (!n) return NaN;
	// Binary search for the first entry above x.
	let lo = 0;
	let hi = n;
	while (lo < hi) {
		const mid = (lo + hi) >> 1;
		if (sorted[mid] <= x) lo = mid + 1; else hi = mid;
	}
	return lo / n;
}

/**
 * The mean of the sample above its `q`-quantile: what the bad cases average to.
 *
 * The number a regulator asks for after the 95th percentile -- not where the
 * tail starts but how heavy it is.
 */
export function conditionalTailExpectation(sorted, q) {
	const n = sorted.length;
	if (!n) return NaN;
	const from = Math.min(n - 1, Math.floor(Math.min(1, Math.max(0, q)) * n));
	let s = 0;
	for (let i = from; i < n; i++) s += sorted[i];
	return s / (n - from);
}

/** The percentiles a summary tabulates. */
export const PERCENTILES = [0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99];

/**
 * The summary itself.
 *
 * @param {Float64Array} sorted  from `sortedColumn`
 * @returns {object} n, mean, sd, skewness, kurtosis (excess), min, max,
 *   meanBounds (95%), percentiles, dkw (the 95% band half-width on the CDF)
 */
export function describeSample(sorted) {
	const n = sorted.length;
	if (!n) {
		return {
			n: 0, mean: NaN, sd: NaN, skewness: NaN, kurtosis: NaN, min: NaN, max: NaN,
			meanBounds: [NaN, NaN], percentiles: [], dkw: NaN,
		};
	}
	let s = 0;
	for (let i = 0; i < n; i++) s += sorted[i];
	const mean = s / n;
	let m2 = 0;
	let m3 = 0;
	let m4 = 0;
	for (let i = 0; i < n; i++) {
		const d = sorted[i] - mean;
		const d2 = d * d;
		m2 += d2;
		m3 += d2 * d;
		m4 += d2 * d2;
	}
	// The sample standard deviation (n−1), which is what a confidence bound
	// on the mean wants; the moments for shape use the population form, which
	// is the textbook definition of the coefficients.
	const sd = n > 1 ? Math.sqrt(m2 / (n - 1)) : 0;
	const v = m2 / n;
	const skewness = v > 0 ? (m3 / n) / v ** 1.5 : NaN;
	const kurtosis = v > 0 ? (m4 / n) / (v * v) - 3 : NaN;
	const half = n > 1 ? 1.959964 * sd / Math.sqrt(n) : 0;
	return {
		n,
		mean,
		sd,
		skewness,
		kurtosis,
		min: sorted[0],
		max: sorted[n - 1],
		meanBounds: [mean - half, mean + half],
		percentiles: PERCENTILES.map((p) => ({ p, value: valueAt(sorted, p) })),
		// DKW: P(sup|F_n − F| > ε) ≤ 2·exp(−2nε²), so at 95% ε = √(ln(2/0.05)/(2n)).
		dkw: Math.sqrt(Math.log(2 / 0.05) / (2 * n)),
	};
}

/** How many bins a histogram may be asked for. */
export const HIST_BINS = { min: 2, max: 200 };

/**
 * A histogram of the sample, for the density panel.
 *
 * Bins by Sturges' rule unless told otherwise, on a log axis where the sample
 * is positive and spans decades -- a dose sample is, and on a linear axis it
 * is one tall bin and a tail nobody can see.
 *
 * `scale` overrules that judgement: `linear` and `log` are the reader's own
 * choice of what the bin *edges* are spaced by, which is a different question
 * from how the counts are drawn and is the one that decides what shape the
 * sample appears to have. **A log request on a sample that reaches zero or
 * below cannot be honoured** -- there is no logarithm of it to bin by -- so it
 * falls back to linear and says so in `wanted`, which is what the panel reads
 * to explain itself.
 *
 * @param {Float64Array} sorted  the column, ascending
 * @param {number|null} [bins]   how many, or null for Sturges' rule
 * @param {'auto'|'linear'|'log'} [scale] what the edges are spaced by
 * @returns {{edges: Float64Array, counts: Uint32Array, log: boolean,
 *   wanted: string, refused: boolean}}
 */
export function histogram(sorted, bins = null, scale = 'auto') {
	const n = sorted.length;
	if (n < 2) {
		return {
			edges: new Float64Array(0), counts: new Uint32Array(0),
			log: false, wanted: scale, refused: false,
		};
	}
	const asked = Math.round(Number(bins));
	const k = Number.isFinite(asked) && asked >= HIST_BINS.min
		? Math.min(HIST_BINS.max, asked)
		: Math.max(5, Math.min(60, Math.ceil(Math.log2(n) + 1)));
	const lo = sorted[0];
	const hi = sorted[n - 1];
	// Decades, judged on the body of the sample and not its extremes: the
	// smallest of two thousand uniforms is a ten-thousandth, and that alone
	// would put a flat sample on a log axis.
	const can = lo > 0;
	const log = scale === 'log' ? can
		: scale === 'linear' ? false
			: can && valueAt(sorted, 0.95) / valueAt(sorted, 0.05) > 100;
	const refused = scale === 'log' && !can;
	const edges = new Float64Array(k + 1);
	for (let b = 0; b <= k; b++) {
		edges[b] = log
			? Math.exp(Math.log(lo) + (Math.log(hi) - Math.log(lo)) * (b / k))
			: lo + (hi - lo) * (b / k);
	}
	const counts = new Uint32Array(k);
	for (let i = 0; i < n; i++) {
		const x = sorted[i];
		let b = log
			? Math.floor(((Math.log(x) - Math.log(lo)) / (Math.log(hi) - Math.log(lo))) * k)
			: Math.floor(((x - lo) / (hi - lo)) * k);
		if (b >= k) b = k - 1;
		if (b < 0) b = 0;
		counts[b]++;
	}
	return { edges, counts, log, wanted: scale, refused };
}

File: kompartment/python/kompartment/stats/distribution.py
"""What a sample of one output looks like, in numbers: Kompartment's distribution summary.

A port of the application's ``src/domain/distribution.js``: one output at one
time over the realisations, sorted once so that every question after the first
is a lookup -- the mean with normal-approximation bounds, standard deviation,
skewness and excess kurtosis, a percentile table, the probability of a value,
the conditional tail expectation, a histogram, and the Dvoretzky-Kiefer-
Wolfowitz band on the whole distribution function, which holds for any shape.

The percentile at ``p`` is the value at position ``p * (n - 1)`` of the sorted
sample, linearly interpolated (a spreadsheet's PERCENTILE). The arithmetic is
the application's, in the same order, with V8's logarithms and exponentials
(see :mod:`.pdf`); the skewness uses the platform's ``pow`` and agrees to the
last bit or two.
"""

from __future__ import annotations

import math
from typing import Any, Dict, List, Optional, Sequence

from .pdf import _div, _exp, _jmax, _jmin, _js_round, _log, _log_array, _pow, _to_number

__all__ = [
    'sorted_column', 'value_at', 'probability_of', 'conditional_tail_expectation', 'PERCENTILES',
    'describe_sample', 'HIST_BINS', 'histogram',
]

#: The percentiles a summary tabulates.
PERCENTILES: List[float] = [0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99]

#: How many bins a histogram may be asked for.
HIST_BINS: Dict[str, int] = {'min': 2, 'max': 200}


def _seq_sum(values: Any) -> float:
    """``s = 0; for (v of values) s += v``: a sum in order, as the application adds."""
    import numpy as np
    if len(values) == 0:
        return 0.0
    with np.errstate(all='ignore'):
        return float(np.cumsum(np.concatenate(([0.0], values)))[-1])


def _typed_sort(values: Any) -> Any:
    """``Float64Array.prototype.sort``: ascending, -0 before +0, NaN last."""
    import numpy as np
    s = np.sort(np.array(values, dtype=np.float64), kind='stable')
    zeros = np.flatnonzero(s == 0)
    if zeros.size:
        negative = int(np.signbit(s[zeros]).sum())
        s[zeros[:negative]] = -0.0
        s[zeros[negative:]] = 0.0
    return s


def sorted_column(values: Sequence[float], times: int, iterations: int, at: int,
                  mask: Optional[Sequence[Any]] = None) -> Any:
    """One output at one time, over the realisations, sorted and finite: a float64 numpy array.

    ``values`` is realisation-major, ``iterations x times``: realisation ``i``
    at time ``j`` is ``values[i * times + j]``; ``at`` is the time. ``mask``,
    where given, keeps realisation ``i`` where ``mask[i]`` is true (non-zero)
    and screens it out otherwise. NaN and infinite values -- realisations that
    failed -- are left out.
    """
    import numpy as np
    vals = np.asarray(values, dtype=np.float64).ravel()
    idx = np.arange(int(iterations), dtype=np.int64) * int(times) + int(at)
    keep = (idx >= 0) & (idx < len(vals))
    if mask is not None:
        m = np.zeros(len(idx), dtype=bool)
        given = np.asarray(mask, dtype=np.float64).ravel()[:len(idx)]
        m[:len(given)] = (given != 0) & ~np.isnan(given)
        keep &= m
    picked = vals[idx[keep]]
    return _typed_sort(picked[np.isfinite(picked)])


def value_at(sorted_: Sequence[float], p: float) -> float:
    """The value at cumulative probability ``p``, interpolated between order statistics."""
    n = len(sorted_)
    if not n:
        return math.nan
    if n == 1:
        return float(sorted_[0])
    x = _jmin(1.0, _jmax(0.0, _to_number(p))) * (n - 1)
    if x != x:
        return math.nan
    lo = math.floor(x)
    hi = min(n - 1, lo + 1)
    a = float(sorted_[lo])
    return a + (float(sorted_[hi]) - a) * (x - lo)


def probability_of(sorted_: Sequence[float], x: float) -> float:
    """The fraction of the sample at or below ``x``: the empirical CDF."""
    n = len(sorted_)
    if not n:
        return math.nan
    x = _to_number(x)
    lo, hi = 0, n
    while lo < hi:
        mid = (lo + hi) >> 1
        if sorted_[mid] <= x:
            lo = mid + 1
        else:
            hi = mid
    return lo / n


def conditional_tail_expectation(sorted_: Sequence[float], q: float) -> float:
    """The mean of the sample above its ``q``-quantile: what the bad cases average to.

    Not where the tail starts but how heavy it is -- the number a regulator
    asks for after the 95th percentile.
    """
    import numpy as np
    n = len(sorted_)
    if not n:
        return math.nan
    t = _jmin(1.0, _jmax(0.0, _to_number(q))) * n
    if t != t:
        return math.nan
    start = min(n - 1, math.floor(t))
    return _seq_sum(np.asarray(sorted_, dtype=np.float64)[start:]) / (n - start)


def describe_sample(sorted_: Sequence[float]) -> Dict[str, Any]:
    """The summary itself, of a column from :func:`sorted_column`.

    Returns ``{'n', 'mean', 'sd', 'skewness', 'kurtosis', 'min', 'max',
    'meanBounds', 'percentiles', 'dkw'}``: the sample SD (over n - 1), the
    population skewness and excess kurtosis, the 95% bounds on the mean, the
    :data:`PERCENTILES` as ``[{'p', 'value'}]``, and ``dkw``, the half-width of
    the 95% Dvoretzky-Kiefer-Wolfowitz band on the CDF, ``sqrt(ln(40) / 2n)``.
    """
    import numpy as np
    a = np.asarray(sorted_, dtype=np.float64)
    n = len(a)
    if not n:
        return {
            'n': 0, 'mean': math.nan, 'sd': math.nan, 'skewness': math.nan, 'kurtosis': math.nan,
            'min': math.nan, 'max': math.nan, 'meanBounds': [math.nan, math.nan], 'percentiles': [],
            'dkw': math.nan,
        }
    mean = _seq_sum(a) / n
    with np.errstate(all='ignore'):
        d = a - mean
        d2 = d * d
        m2 = _seq_sum(d2)
        m3 = _seq_sum(d2 * d)
        m4 = _seq_sum(d2 * d2)
    # The sample standard deviation (n - 1), which a bound on the mean wants;
    # the population form for the shape.
    sd = math.sqrt(m2 / (n - 1)) if n > 1 else 0
    v = m2 / n
    skewness = _div(m3 / n, _pow(v, 1.5)) if v > 0 else math.nan
    kurtosis = _div(m4 / n, v * v) - 3 if v > 0 else math.nan
    half = 1.959964 * sd / math.sqrt(n) if n > 1 else 0
    return {
        'n': n,
        'mean': mean,
        'sd': sd,
        'skewness': skewness,
        'kurtosis': kurtosis,
        'min': float(a[0]),
        'max': float(a[n - 1]),
        'meanBounds': [mean - half, mean + half],
        'percentiles': [{'p': p, 'value': value_at(a, p)} for p in PERCENTILES],
        # DKW: P(sup|F_n - F| > e) <= 2 exp(-2 n e**2), so at 95% e = sqrt(ln(2/0.05) / 2n).
        'dkw': math.sqrt(_log(2 / 0.05) / (2 * n)),
    }


def histogram(sorted_: Sequence[float], bins: Any = None, scale: str = 'auto') -> Dict[str, Any]:
    """A histogram of an ascending sample, for the density panel.

    ``bins`` is how many (clamped to :data:`HIST_BINS`; ``None`` or fewer than
    two for Sturges' rule, 5 to 60). ``scale`` is what the bin edges are spaced
    by: ``'linear'``, ``'log'``, or ``'auto'`` -- a log axis where the sample
    is positive and its 5th to 95th percentiles span more than two decades. A
    log request on a sample that reaches zero or below cannot be honoured; it
    falls back to linear and says so in ``refused``.

    Returns ``{'edges', 'counts', 'log', 'wanted', 'refused'}``: ``edges`` a
    float64 array one longer than ``counts`` (uint32), ``wanted`` the scale
    asked for. A sample of fewer than two has no histogram (empty arrays).
    """
    import numpy as np
    a = np.asarray(sorted_, dtype=np.float64)
    n = len(a)
    if n < 2:
        return {'edges': np.zeros(0, dtype=np.float64), 'counts': np.zeros(0, dtype=np.uint32),
                'log': False, 'wanted': scale, 'refused': False}
    asked = _js_round(_to_number(bins))
    if math.isfinite(asked) and asked >= HIST_BINS['min']:
        k = int(min(HIST_BINS['max'], asked))
    else:
        k = max(5, min(60, math.ceil(math.log2(n) + 1)))
    lo = float(a[0])
    hi = float(a[n - 1])
    # Decades, judged on the body of the sample and not its extremes.
    can = lo > 0
    if scale == 'log':
        log = can
    elif scale == 'linear':
        log = False
    else:
        log = can and value_at(a, 0.95) / value_at(a, 0.05) > 100
    refused = scale == 'log' and not can
    edges = np.empty(k + 1, dtype=np.float64)
    for b in range(k + 1):
        if log:
            edges[b] = _exp(_log(lo) + (_log(hi) - _log(lo)) * (b / k))
        else:
            edges[b] = lo + (hi - lo) * (b / k)
    with np.errstate(all='ignore'):
        if log:
            where = np.floor(((_log_array(a) - _log(lo)) / (_log(hi) - _log(lo))) * k)
        else:
            where = np.floor(((a - lo) / (hi - lo)) * k)
    # A bin that is not a number -- a sample all of one value, 0/0 -- counts
    # nowhere, as the application's typed array ignores such an index.
    counted = ~np.isnan(where)
    where = np.where(where >= k, k - 1, where)
    where = np.where(where < 0, 0, where)
    counts = np.bincount(where[counted].astype(np.int64), minlength=k).astype(np.uint32)
    return {'edges': edges, 'counts': counts, 'log': bool(log), 'wanted': scale, 'refused': bool(refused)}
